numpy/core/src/multiarray/arrayfunction_override.c

```diff
@@ -329,6 +336,107 @@ array_implement_array_function(
 }
 
 
+/*
+ * Implements the __array_function__ protocol for a Python function, as described in
+ * in NEP-18. See numpy.core.overrides for a full docstring.
+ */
+NPY_NO_EXPORT PyObject *
+array_implement_array_function(
+    PyObject *NPY_UNUSED(dummy), PyObject *positional_args)
+{
+    PyObject *implementation, *public_api, *relevant_args, *args, *kwargs;
+
+    if (!intern_strings()) {
+        if (PyErr_Occurred()) {
+            PyErr_SetString(PyExc_RuntimeError,
+                            "cannot load arrayfunction_override.");
+        }
+        return NULL;
+    }
+
+    if (!PyArg_UnpackTuple(
+            positional_args, "implement_array_function", 5, 5,
+            &implementation, &public_api, &relevant_args, &args, &kwargs)) {
+        return NULL;
+    }
+
+    /* Remove `like=` kwarg, which is NumPy-exclusive and thus not present
+     * in downstream libraries.
+     */
+    if (PyDict_CheckExact(kwargs) &&
+            PyDict_Contains(kwargs, npy_arrayfunction_str_like)) {
+        PyDict_DelItem(kwargs, npy_arrayfunction_str_like);
+    }
+
+    return array_implement_array_function_internal(
+        implementation, public_api, relevant_args, args, kwargs);
+}
+
+
+/*
+ * Implements the __array_function__ protocol for a C function. Added as an extension
+ * to NEP-18 in an effort to bring NEP-35 to life with minimal dispatch overhead.
+ */
+NPY_NO_EXPORT PyObject *
+array_implement_c_array_function(
+    const char *function_name, PyObject *args, PyObject *kwargs,
+    PyObject **err_type, PyObject **err_value, PyObject **err_traceback)
+{
+    PyObject *relevant_args;
+    PyObject *numpy_module = NULL, *public_api = NULL;
+    PyObject *result = NULL;
+
+    if (kwargs == NULL)
+        return NULL;
+
+    if (!intern_strings()) {
+        if (PyErr_Occurred()) {
+            PyErr_SetString(PyExc_RuntimeError,
+                            "cannot load arrayfunction_override.");
+        }
+        goto cleanup;
+    }
+
+    /* Remove `like=` kwarg, which is NumPy-exclusive and thus not present
+     * in downstream libraries. If that key isn't present, return NULL and
+     * let originating call to continue.
+     */
+    if (PyDict_CheckExact(kwargs) &&
+            PyDict_Contains(kwargs, npy_arrayfunction_str_like)) {
+        relevant_args = PyTuple_Pack(1,
+                PyDict_GetItem(kwargs, npy_arrayfunction_str_like));
+        PyDict_DelItem(kwargs, npy_arrayfunction_str_like);
+    }
+    else {
+        return NULL;
+    }
+
+    numpy_module = PyImport_Import(npy_arrayfunction_str_numpy);
+    if (numpy_module != NULL) {
+        public_api = PyObject_GetAttrString(numpy_module, function_name);
+        if (public_api == NULL || !PyCallable_Check(public_api)) {
+            if (PyErr_Occurred()) {
+                PyErr_Format(PyExc_RuntimeError,
+                             "cannot import numpy.%s or it is not callable.",
```

Comment:
You should separate this into two ifs. If it can't be imported, the exception has already been set for you

Reply:
Done.